GPII-3117: Remove the EnvironmentReporter and adjust ContextManager accordingly

klown opened the pull request:
JIRA: https://issues.gpii.net/browse/GPII-3117

- Removed periodic invocation of EnvironmentReporter from ContextManager,
- Removed PUT /environmentChanged endpoint,
- Removed support for `conditions:` blocks in prefsSafes, and removed the blocks themselves from preferences data and test code,
- Moved the remaining functionality to the LifeCycleManager, and removed the ContextManager entirely,
- Continued support for manual context (now "prefsSets") changes by the user via a preferences editor, via PSPChannel integration.

NOTE: Before merging with master, tag master with 'EnvironmentReporter' to easily find the removed functionality if we decide to go back to it.

cindyli commented:
There's a PSPChannel integration test that tests the PSPChannel context change API with a `/environmentChanged` request - https://github.com/GPII/universal/blob/master/tests/shared/PSPIntegrationTestDefs.js#L25-L29

This test can be removed too.

klown commented:
Thanks, @cindyli.  I'll remove it.

What's puzzling is there is no /environment changed endpoint anymore, so why does the test pass?  Specifically, why is there no `404` or other error code for that request?  The log shows:

> 15:59:00.955:  Sending a PUT request to: /environmentChanged on port 8081
> 15:59:00.982:  Kettle server getDispatcher found no matching handlers, continuing

Hopefully there aren't any more of these lurking around.

klown commented:
Ready for review @cindyli although the `README` needs work.

Also, I found it tricky to make one of the integration tests work.  I came up with a solution that involves storing context information in the `ContextManager`.  I'm not sure that it's the correct way to go.  The test case is possibly invalid now.  The test case is where [the context changes before a user logs in](https://github.com/klown/universal/blob/GPII-3117/tests/shared/ContextIntegrationTestDefs.js#L250).  Previously, the `ContextManager` stored information about the environment such that when a user logged in, the context appropriate to the stored environment was available.  Removing all the code/data for the environment means that context information is gone, and the test fails.  Storing the context name (e.g., "bright"), in the ContextManager model is sufficient to make the test work again.  In addition, making that change has not caused regressions elsewhere.

Still, I don't know enough to say whether storing a context in the ContextManager is useful or legitimate.  Do you have any insight?

The same technique is used in one of the [PSPIntegrationTests](https://github.com/klown/universal/blob/GPII-3117/tests/shared/PSPIntegrationTestDefs.js#L589) that previously depended on a `/environmentChanged` request.  But in this case, the change in context should not change the settings (and it doesn't).  I could see removing this test entirely, but, again, I'm not completely sure.

amb26 commented:
Doesn't make sense - if the EnvironmentReporter is removed, we should remove the ContextManager implementation and its tests also since they serve no user function.

cindyli commented:
@amb26, the code left for ContextManager in @klown's ContextManager serves as a bridge to support the PSPChannel context change API  that allows the switch from for example "gpii-default" to "bright". The remaining code sets `lifecycleManager.session.model.activeContextName` to the new context name then triggers `lifecycleManager.update()`. If it's OK to add a new `contextChange()` function into lifecycleManager and get PSPChannel trigger it directly, ContextManager can be removed completely. 

Is this what you meant by removing the ContextManager implementation?

cindyli commented:
@klown, [the PSPIntegrationTests you mentioned](https://github.com/klown/universal/blob/GPII-3117/tests/shared/PSPIntegrationTestDefs.js#L589) can be removed completely.

klown commented:
@amb26 , as @cindyli noted, the point was to keep the functionality that allowed context changes coming through the PSPChannel.  The choice is either keep that contained inside a ContextManager component, or disperse the functionality elsewhere.

FWIW, I've found it easier to concentrate on a component called "ContextManager" and not think about where the functionality and tests would migrate to if there was not such component.

klown commented:
@cindyli 

> @klown, the PSPIntegrationTests you mentioned can be removed completely.

Yes, it doesn't look like it was doing anything real anymore, but I wasn't sure.

What about the "Context changed before login" test in `[ContextIntegrationTests](https://github.com/klown/universal/blob/GPII-3117/tests/shared/ContextIntegrationTestDefs.js#L250)`?  Towards an answer to the question:
- Can a context change come through the PSPChannel before a user logs in?
- Is that context asserted (stored, noted) such that it has an effect after the user logs in?

cindyli commented:
@klown, I think whether or not removing ContextIntegrationTests depends on the decision of whether ContextManager should be removed.

>     * Can a context change come through the PSPChannel before a user logs in?
A special GPII key "noUser" will log in when there's no real user logs in. It's nice to test a context change doesn't break the system when "noUser" is logged in or even before any user including "noUser" logs in.

>     * Is that context asserted (stored, noted) such that it has an effect after the user logs in?
It may not be a good idea to have ContextManager save the current context even if we decide to keep it since the active context is tracked by [`lifecycleManager.userSession.model.activeContextName`](https://github.com/klown/universal/blob/GPII-3117/gpii/node_modules/contextManager/src/ContextManager.js#L143). Saving the active context in ContextManager will have to ensure the sync-up of this value in 2 components.

amb26 commented:
"Contexts" have long since been a cursed term in the project, since they have ceased to have the meaning of what we now call "preference sets" for over 3 years. @cindyli 's comment of a few minutes ago is spot on, there is no need for separate tracking anywhere else of the piece of state held at lifecycleManager.userSession.model.activeContextName , which we should take the opportunity to rename as activePrefsSetName . As a component the ContextManager itself can be removed wholesale together with its dedicated tests, although as noted we still need tests which verify the overall functionality of changing prefs set name, e.g. via the PCP API. There is no meaning of a "context change before a user logs in" because the only current meaning corresponding to a "context" is a selection from the set of prefs set names which are found to be available in the user's preference safe at the point they log in.

klown commented:
> It may not be a good idea to have ContextManager save the current context even if we decide to keep it since the active context is tracked by `lifecycleManager.userSession.model.activeContextName`. Saving the active context in ContextManager will have to ensure the sync-up of this value in 2 components.

The way it's written keeps them in sync, with one exception.  Glossing over details, if `ContextManager.updateCurrentContext()` is called with new context:
- it updates the ContextManager's model, causing a model change
- that triggers the ContextManager's `contextChanged()` model listener
- that updates the `lifecycleManager.userSession.model.activeContextName`

Note that this works the same way it did when the ContextManager model was a `currentEnvironment`.  The difference is storing a context vs. an environment.

The one exception is when no one is logged in (more accurately when a reserved GPII key is active).  That relates to the test in question -- changing the context when no one is logged in.  Based on the logic of the code, I'm guessing it's undesirable to store a pre-login context name in the `lifecycleManager.userSession.model.activeContextName`, but I don't know why. 

What is problematic about storing a change in context in the LifecycleManager when no one is logged in?  Perhaps it isn't.  If so, that would be another reason to get rid of the ContextManager.

klown commented:
Ready for review, @cindyli @amb26

cindyli commented on documentation/LoginAndLogoutFlow.md:
```diff
@@ -75,8 +75,8 @@ The user login process is as follows:
    details of this flow except for some general observation:
-   1. Generally some prioritized steps: `preProcess`, `matchMakerDispatcher`, `runContextManager` and `transform` will
-      be run (in that order).
+   1. Generally some prioritized steps: `preProcess`, `matchMakerDispatcher`, `updsteActivePrefsSetName` and `transform`
```
typo. updste -> update?

klown replied:
and there's another one in `documentation/CloudBasedFlow.md`

cindyli commented on gpii/node_modules/lifecycleManager/src/LifecycleManager.js:
```diff
@@ -1309,4 +1319,94 @@ var gpii = fluid.registerNamespace("gpii");
+     *    configuration is returned. If the call to prefsSetChanged does not result in a different
+     *    active preferences set or no user is logged in, undefined is returned.
+     **/
+    gpii.lifecycleManager.prefsSetNameChanged = function (that, forcePrefsSetName) {
```
As now we only need to support the context change request from PSPChannel, the logic of this function can be cleaned up and simplified:

1. start with verifications:
1.1 if no one is logged in, or, a system reserved key is logged in, do nothing
1.2 if the new prefs set is not provided (`!newPrefsSetName`) or it's same as the active prefs set (`activeSession.model.activePrefsSetName === newPrefsSetName`), do nothing

2. update the payload: set prefs set name; delete appliedSolutions; calculate the new lifecycle instruction. Also, It's no longer necessary to keep `gpii.lifecycleManager.updateActivePrefsSetName()` function. It can be replaced by one single assignment - `fullPayload.activePrefsSetName = newPrefsSetName`

3. apply updated payload: run `lifecycleManager.update(payload)` and update activeSession.model in its promise callback.

Please also clean up comments that no longer make sense.

klown replied:
> 1.2 if the new prefs set is not provided (!newPrefsSetName) or it's same as the active prefs set (activeSession.model.activePrefsSetName === newPrefsSetName), do nothing

What is `newPrefsSetName`?  Do you mean the local`newActivePrefsSetName`?  Or, do you mean to replace `forcePrefsSetName` with `newPrefsSetName`?

Also, there's a difference between this new logic and the old in that the old logic could result in a new prefs set name of `gpii-default`.  In the old logic, this happened when:
- no new name is passed in (`forcePrefsSetName` is undefined)
- `activeSession.model.activePrefsSetName` is undefined or null

See the beginning of [`gpii.lifecycleManager.updateActivePrefsSetName()`](https://github.com/klown/universal/blob/GPII-3117/gpii/node_modules/lifecycleManager/src/LifecycleManager.js#L1332), where the new name is assigned a default of `gpii-default`. Is a default new name of `gpii-default` still a possibility?

cindyli replied:
>What is newPrefsSetName? Do you mean the localnewActivePrefsSetName? Or, do you mean to replace forcePrefsSetName with newPrefsSetName?
The `newPrefsSetName` I meant was the new prefsSetName that comes in from PSPChannel requests.

In terms of the variable name, "newPrefsSetName" seems better than "forcePrefsSetName".

>there's a difference between this new logic and the old in that the old logic could result in a new prefs set name of gpii-default.
Yes, the default pref set name is "gpii-default". I believe the old logic that auto-switches to "gpii-default" was created to handle the case when the environment change doesn't fall into any pref set. In the new logic, it would be cleaner to have the new function expect a valid new prefs set name and simply switch to it.

klown replied:
>  Also, It's no longer necessary to keep gpii.lifecycleManager.updateActivePrefsSetName() function.

I just removed it and tests are failing.  It's used (at least) by the `gpii.lifecycleManager.userLogonHandling.stateChangeHandler` on a `onMatchDone` event.  See https://github.com/klown/universal/blob/GPII-3117/gpii/node_modules/lifecycleManager/src/UserLogonStateChange.js#L206.  I'm looking into how to handle this, but if you have any insights, that would be helpful.

cindyli replied:
Good catch, @klown. ya, seems we need to keep this function hopefully with a cleaner logic.

cindyli commented on gpii/node_modules/lifecycleManager/src/LifecycleManager.js:
```diff
@@ -1309,4 +1319,94 @@ var gpii = fluid.registerNamespace("gpii");
+            // "finalPayload" as instructions on how to configure the system. once configured,
+            // the applied settings are stored in 'activeConfiguration'
+            delete fullPayload.lifecycleInstructions;
+            // TODO: This is pretty rubbish, why isn't the natural action of lifecycleManager.update good enough here?
```
This comment should be removed. start() and update() only apply settings to solutions but don't touch session model.

klown replied:
So, the commenter's question has been answered.  I wonder if they know...

Sure, I'll remove it.

cindyli commented on gpii/node_modules/lifecycleManager/src/LifecycleManager.js:
```diff
@@ -1309,4 +1319,94 @@ var gpii = fluid.registerNamespace("gpii");
+            // to the sessionState in the next line. LifecycleInstructions is only used in the
+            // "finalPayload" as instructions on how to configure the system. once configured,
+            // the applied settings are stored in 'activeConfiguration'
+            delete fullPayload.lifecycleInstructions;
```
I'm not sure if this deletion ever makes any sense since it deletes `lifecycleInstructions` at the wrong path. "lifecycleInstructions" is located at `fullPayload.activeConfiguration.lifecycleInstructions`. Missing `activeConfiguration`.

klown replied:
Well, I have no idea, and it makes me nervous that you are "not sure" :-).  Should this `delete` be removed entirely?  Or should the correct path be used?  I'll look around and see if I can make sense of this.

cindyli replied:
`lifecycleInstructions` was not removed from the session model when issuing model change requests at other places in lifecycleManager. But as you said it's always helpful to double check if the same logic applies here. Thanks.

cindyli commented on gpii/node_modules/lifecycleManager/src/LifecycleManager.js:
```diff
@@ -532,7 +542,7 @@ var gpii = fluid.registerNamespace("gpii");
     gpii.lifecycleManager.applyWithNewPrefs = function (that, session, contextualPrefs, prefsAppliedLocalEvent) {
         var newPrefsSet = {};
-        fluid.set(newPrefsSet, ["contexts", session.model.activeContextName, "preferences"], contextualPrefs);
+        fluid.set(newPrefsSet, ["contexts", session.model.activePrefsSetName, "preferences"], contextualPrefs);
```
It's odd to see expressions like `["contexts", session.model.activePrefsSetName, "preferences"]` where one name has been renamed to use "prefsSet" as `activePrefsSetName` while its parent is still named as `contexts`. This situation repeatedly shows up in the code base.

To sync up all names, the ideal way is to rename all "context" to "prefsSet". But doing so will need a database upgrade to make this change for all preference safe documents in CouchDB.

Or we stick with the old name of "context".

What do you think, @amb26?

cindyli commented:
As contextManager tests have been removed, it's the time to enhance PSPChannel integration tests for prefs set change at [here](https://github.com/klown/universal/blob/GPII-3117/tests/shared/PSPIntegrationTestDefs.js#L581-L583) as well as line 588-590, 640-642, 660-662 to check not only the message type, but also "settingControls" received by PSPChannel client to make sure preferences in the new prefs set are received.

klown commented:
@cindyli I think I've got the logic correct for LifecycleManger's `updateActivePrefsSetName()` and `prefsSetNameChanged()` functions.  And, I've extended the integration tests as you suggested (and even added a few more).

Ready for review.

cindyli commented on testData/preferences/manuel.json5:
```diff
@@ -61,14 +61,6 @@
                         "value": 1024
                     }
-                ],
-                "conditions": [
```
The comment at line 17, 18 explaining the condition change should be removed. Thanks.

klown replied:
Done

cindyli commented:
`documentation/ContextManager.md` should be removed. If you search the code base using the keyword `conditions`, it appears in a couple of .md, .json5 and .js files. Please adjust accordingly. Thanks.

cindyli commented:
Removed ContextManager files should be removed from [.nycrc](https://github.com/klown/universal/blob/GPII-3117/.nycrc#L9-L11).

Searching the code base with `activeContext` shows 12 matches in a few files. Please consider to change them to `activePrefsSet`.

cindyli commented on gpii/node_modules/lifecycleManager/src/LifecycleManager.js:
```diff
@@ -1309,4 +1319,85 @@ var gpii = fluid.registerNamespace("gpii");
+     */
+    gpii.lifecycleManager.updateActivePrefsSetName = function (that, fullPayload, newPrefsSetName) {
+        var activeSessionPrefsSetName = that.getSession().model.activePrefsSetName;
+        if (!newPrefsSetName) {
```
line 1337-1344 can be simplified to:

```
fullPayload.activePrefsSetName = newPrefsSetName ? newPrefsSetName : activeSessionPrefsSetName ? activeSessionPrefsSetName : "gpii-default";
```

cindyli commented on tests/shared/ContextIntegrationTestDefs.js:
```diff
@@ -24,59 +24,69 @@ fluid.registerNamespace("gpii.tests.contextIntegration");
-
 // NOTE: This inherits from from "gpii.test.testCaseHolder" via "gpii.test.integration.testCaseHolder.linux"
 // from which it gets loginRequest, logoutRequest and other standard events
 fluid.defaults("gpii.tests.contextIntegration.testCaseHolder", {
```
With the removal of ContextManager, this file can be removed too.

klown replied:
> Removed ContextManager files should be removed from [.nycrc](https://github.com/klown/universal/blob/GPII-3117/.nycrc#L9-L11).

Okay, done.
> 
> Searching the code base with `activeContext` shows 12 matches in a few files. Please consider to change them to `activePrefsSet`.

I could only find three matches for "activeContext" (singular), but maybe that's because I had already removed files or sections of files with "conditions" or "ContextManger", etc.  The three remaining matches were all relative to the `MatchMaker`.  Here are the links, where `activeContexts` has been replaced with `activePrefsSets`:

- [matchmaker.output.payload.json](https://github.com/klown/universal/blob/GPII-3117/gpii/node_modules/flowManager/test/data/matchmaker.output.payload.json#L2)
- [MatchMakerUtilitiesTests.js](https://github.com/klown/universal/blob/GPII-3117/gpii/node_modules/matchMakerFramework/test/js/MatchMakerUtilitiesTests.js#L345)
- [MatchMakerUtilities.js](https://github.com/klown/universal/blob/GPII-3117/gpii/node_modules/matchMakerFramework/src/MatchMakerUtilities.js#L50)

I'm still pondering that last one because, so far, I can't see where it's used.

klown replied:
@cindyli I think I've found a bug, and I looking close to confirm.  Here's an summary of the "problem".  Note that the links below are to the current master branch.

The MatchMaker utility `gpii.matchMakerFramework.utils.preProcess()` allocates an  [`activeContexts[]`](https://github.com/GPII/universal/blob/master/gpii/node_modules/matchMakerFramework/src/MatchMakerUtilities.js#L50) array.

Later the cloud-based FlowManager's SettingsGetHandler performs the match in `gpii.flowManager.cloudBased.matchToSettings()`.  The first operation there is to filter the `finalPayload` in terms of keys such as `gpiiKey`, `preferences`, and others.  In particular, it looks for the `activeContextNames` key.  Note the difference between `activeContexts` and `activeContextNames`.  I don't think the SettingsGetHandler ever retrieves the `activeContexts`.

However, I've put in some log messages, and am running all of the tests to see what `matchToSettings()` actually does. I'll let you know the results.

klown commented:
@cindyli 
I wrote:

> I've put in some log messages, and am running all of the tests to see what `matchToSettings()` actually does. I'll let you know the results.

The results of running all the node tests with this extra logging information is listed below.  The input to the match is `finalPayload` which is filtered in creating `settings`.  After the filtering, `settings.preferences` is  further processed by `gpii.matchMakerFramework.utils.filterPreferencesFromInferredConfig()`.

- `finalPayload.activeContexts` exists, but is always empty
- `finalPayload.activeContextName` does not exist
-  neither `settings.activeContexts` nor `settings.activeContextName` ever exist.

I've also removed the `activeContexts` array where it is first added, and re-run the tests.  Nothing fails.  So far, it looks like `activeContexts` is superfluous in this context (pardon the pun).

The only thing left to check are the browser tests to see if `activeContexts` shows up there.

klown commented:
@cindyli  @amb26  I've removed the use of `activePrefsSets` (formerly `activeContexts`) local to [`MatchMakerUtilities.js`](https://github.com/klown/universal/blob/GPII-3117/gpii/node_modules/matchMakerFramework/src/MatchMakerUtilities.js#L48).   I cannot find its use anywhere.  For the same reason, I've deleted the test data file [`matchmaker.out.payload.json`](https://github.com/klown/universal/commit/a89fb9ed6513e5726fe1240b5b4827a87f621af1#diff-7696c30a21ca4f14c4168df6cb31f071).

If you know that these are useful somehow, let me know and I'll put them back.  Otherwise, this is ready for another review, thanks.

cindyli commented:
In "documentation/README.md", line 16 can be removed. It points to "context manager" doc that no longer exists.

"context" appears in "documentation/MatchMakerFramework.md" at line 5 and 53 should be changed to "preferences set".

In "documentation/PreferencesServer.md" line 12, "according to a specific set of contexts" -> "according to a specific preferences set".

In "documentation/SolutionRegistryFormat.md" line 311, "context changes" -> "preferences set changes".

In "documentation/Transformer.md" line 13, remove "/context manager".

In "examples/pspChannelClient/README.md", modify various appearance of "activeContextName" and "context".

In "examples/pspChannelClient/pspChannelClient.js" line 40, "activeContextName" -> "activePrefsSetName".

In "gpii/node_modules/eventLog/src/metrics.js", "gpii.metrics.trackContextChange" function name should be renamed to "gpii.metrics.trackPrefsSetChange". Also at line 174, 175, "active context" -> "active preferences set".

In "gpii/node_modules/flowManager/src/PSPChannel.js" and ,"gpii/node_modules/flowManager/test/PSPChannelTests.js", except the use of "contexts", most appearance of "context" should be changed to "preferences set".

Please also check the use of "context" in "LifecycleManagerSession.js", "MatchMakerUtilities.js", "MatchMakerFrameworkTests.js", "Transformer.js" etc. In short, if you search the keyword "context" in the code base, many use of it should be reconsidered.

cindyli commented on gpii/node_modules/lifecycleManager/src/LifecycleManager.js:
```diff
@@ -1309,4 +1319,81 @@ var gpii = fluid.registerNamespace("gpii");
+            fluid.log("lifecycleManager: Same preferences set as before (" + newPrefsSetName + "): no prefs set change");
+            return;
+        }
+        // Update the prefs set name immediately so we don't produce a model update from the lifecycleManager with a
```
This comment is for line 1379. Please move it one line below.

cindyli commented:
With the removal of "resetWithEnvReportTests", these config files used by the tests can be removed too:

* tests/configs/gpii.tests.acceptance.resetWithEnvReport.config.json5
* tests/configs/gpii.tests.acceptance.untrusted.resetWithEnvReport.config.json5

klown commented:
@cindyli 
I think I've replaced or removed all occurrences of "context manager" and variants thereof (ContextManager, contextManager, contextmanager, context manager, context_manager, etc) from universal, as well as many cases of "context".

> In short, if you search the keyword "context" in the code base, many uses of it should be reconsidered.

I searched for all the instances of the string `context` in the code base in `universal/gpii/node_modules/...`.  After filtering out:
- `"contexts":`,
- `contexts`,
-  `fluid.contextAware`, and
- `contextValue` (assumed that it's associated with `fluid.contextAware`)

there are still 210 places where "context" is used.  I started to evaluate them but it's not obvious what "context" means in each case -- it might be "context" in the sense of the `ContextManager`, or it might be something else.  It's going to take a while to work through all of these.  In fact, isn't this one of the tasks for [GPII-4228](https://issues.gpii.net/browse/GPII-4228):  'Replace "contexts:" from preferences and codebase' ?

FYI here is the command I used to find potential replacements for "context".  If you know of something that would cut the size down even more, or give more accurate results, let me know:
`grep -r -i context gpii/* | grep -v '\"contexts\":' | grep -v 'contexts:' | grep -v 'fluid.contextAware' | grep -v contextValue | grep -v 'context1'`

cindyli commented:
@klown, your grep command is much smarter than my method of searching "context" in IDE and going through the search result. ;)

I agree the replacement of "contexts: " should be deferred later so we can fix the backend DB data and the code base together. But as we agreed to replace "context -> prefs set" when applicable, I think it's worthwhile to clean up the code base on this part.

You've addressed the majority of the replacement. The last a couple:

1. gpii/node_modules/matchMakerFramework/src/MatchMakerUtilities.js
2. gpii/node_modules/ontologyHandler/src/ontologyHandler.js
3. Let's take the chance to fix [a historical typo](https://github.com/klown/universal/blob/GPII-3117/tests/JournalIntegrationTests.js#L2): Context Integration Tests -> Journal Integration Tests

klown commented:
@cindyli 

> You've addressed the majority of the replacement. The last a couple:
> ...

Okay, I took care of those.

New grep command to ignore "deviceContext":
`grep -r -i context gpii/* | grep -v '\"contexts\":' | grep -v 'contexts:' | grep -v 'fluid.contextAware' | grep -v contextValue | grep -i -v 'ContextAwareness' | grep -v 'context1' | grep -i -v 'deviceContext'`

klown commented:
@cindyli I've double checked universal's `tests` and `testData` sub-folders using my grep command, and investigated the occurrences of "context" it output.  I did not see anything that needed changing to "prefsSet" or similar.
